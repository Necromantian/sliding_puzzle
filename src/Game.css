.game {
    position: relative;
    overflow: hidden;
    height:100vh;
    width:100wh;
  }
  
.boardContainer{
    
    height: 502px;
    width: 502px;
    position: absolute;
    left: 50%;
    top: 50%;
    transform: translate(-50%,-50%);
    
}