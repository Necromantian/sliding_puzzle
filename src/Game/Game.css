.boardContainer{
    border: solid gray 1px;
    height: 500px;
    width: 500px;
    background-color: white;
    position: absolute;
    left: 50%;
    top: 50%;
    transform: translate(-50%,-50%);
}

.board{
    position: relative;
}

.tile{
    height:125px;
    width:125px;
    border: solid black 1px;
    background-color: rgb(46, 141, 46);
}

.test{
    position: absolute;
    bottom:200px;
}
.test2{
    position: absolute;
    bottom:150px;
}

.test3{
    position: absolute;
    bottom:125px;
}

.test4{
    position: absolute;
    bottom:175px;
}