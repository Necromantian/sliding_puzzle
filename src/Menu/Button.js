import React, { Component } from 'react';
import './Menu.css';

class Button extends Component {
  render() {
    return (
      <div className='button'>

      </div>
    );
  }
}

export default Button;