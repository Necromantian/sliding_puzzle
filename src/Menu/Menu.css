.menu{
    border: solid var(--menu-border-color) 2px;
    border-top-left-radius: 15px;
    border-bottom-left-radius: 15px;

    height: 180px;
    width: 100px;
    background-color: var(--menu-background-color);
    position: absolute;
    right:0;
    top: 50%;
    transform: translate(0,-50%);
}

.menu__button{
    display: inline-block;
    padding: 12px;
    margin: 10px;

    background-color: var(--menu-button-color); /* Green */
    border: black 2px solid;
    border-radius: 10px;

    color: var(--main-font-color);
    text-align: center;
    text-decoration: none;
    font-size: 16px;
    font-weight:bold;

    -webkit-transition-duration: 0.4s; /* Safari */
    transition-duration: 0.4s;
}

.menu__button--clicked{
    background-color: white; /* Green */
    color: var(--menu-background-color);
    
}

.menu__button:hover{
    box-shadow: 0 12px 16px 0 rgba(0,0,0,0.24), 0 17px 50px 0 rgba(0,0,0,0.19);
    cursor: pointer;
}

button:active{
    background-color: white;
    color: var(--main-color);
}