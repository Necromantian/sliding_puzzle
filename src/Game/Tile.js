import React, { Component } from 'react';
import './Game.css';

import { Motion, spring } from "react-motion";

class Tile extends Component {

    //fast
    //{stiffness: 900, damping: 40}

    //slow
    //{stiffness: 180, damping: 30}
  render() {
    return (
        <Motion style={{x: spring(this.props.positionX,{stiffness: 180, damping: 30}),y: spring(this.props.positionY)}}>
        {({x, y}) =>
          // children is a callback which should accept the current value of
          // `style`
          <div style={{
            WebkitTransform: `translate3d(${x}px, ${y}px, 0)`,
            transform: `translate3d(${x}px, ${y}px, 0)`,
          }}>
            <div className='tile'></div>
          </div>
        }
      </Motion>

    );
  }
}

export default Tile;