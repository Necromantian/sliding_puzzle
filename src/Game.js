import React, { Component } from 'react';
import './Game.css';

import Board from './Board/Board.js'
import Menu from './Menu/Menu.js'

class Game extends Component {
  constructor(props) {
    super(props);
    this.state = {
      posX: 0,
      posY: 0,
      tilesData: [],
      animationSpeed: "slow",
      changingColors: false
    };
  };

  componentDidMount() {
    let tilesData = this.getInitTilesData();
    this.setState({ tilesData });
  }


  getInitTilesData() {
    let tilesData = [];
    for (let i = 0; i < 15; i++) {
      tilesData.push({ number: i, currentPosition: i });
    }
    return tilesData;
  }

  moveTile(clickedTile, currentPosition) {
    this.setState({ animationSpeed: "fast" }, () => {
      let tilesData = [...this.state.tilesData];

      let possibleMovements = this.getPossibleMovements(currentPosition);
      for (let i = 0; i < possibleMovements.length; i++) {
        let move = possibleMovements[i];
        if (!tilesData.some(t => t.currentPosition === move)) { //if there is no tile with currentPosition on possibleMovements[i]
          tilesData[clickedTile].currentPosition = move;
          this.setState({ tilesData },() => {
            if(this.checkIfWin()){
              alert("WIN")
            }
          });
        }
      }
    });
  }

  getPossibleMovements(currentPosition) {
    let possibleMovements = [];
    if (currentPosition + 1 <= 15 && currentPosition !== 7 && currentPosition !== 11 && currentPosition !== 3) //7,11,3 could go to next row by accident
      possibleMovements.push(currentPosition + 1);

    if (currentPosition + 4 <= 15)
      possibleMovements.push(currentPosition + 4);

    if (currentPosition - 1 >= 0 && currentPosition !== 12 && currentPosition !== 8 && currentPosition !== 4) //7,10,3 could go to previous row by accident
      possibleMovements.push(currentPosition - 1);

    if (currentPosition - 4 >= 0)
      possibleMovements.push(currentPosition - 4);

    return possibleMovements;
  }

  swapTwoRandomTiles = () => {
    let tilesData = [...this.state.tilesData];
    let rand1 = Math.floor(Math.random() * 15);
    let rand2 = Math.floor(Math.random() * 15);

    let temp = tilesData[rand1].currentPosition;
    tilesData[rand1].currentPosition = tilesData[rand2].currentPosition;
    tilesData[rand2].currentPosition = temp;

    this.setState({ tilesData });
  }

  shuffleClick = () => {
    this.setState({ animationSpeed: "slow" }, () => {
      for (let i = 0; i < 50; i++) {
        this.swapTwoRandomTiles();
      }
    });
  }

  changingColorsClick = () => {
    this.setState({ changingColors: !this.state.changingColors })
  }

  checkIfWin() {
    let tilesData = [...this.state.tilesData];

    for (let i = 0; i < tilesData.length; i++) {
      if(tilesData[i].currentPosition !== i){
        return false;
      }
    }
    return true;
  }

  render() {
    return (
      <div className="game">
        <div className='boardContainer'>
          {/* <button onClick={this.shuffle.bind(this)}>Tasuj</button> */}
          <div className="board">
            <Board changingColors={this.state.changingColors} animationSpeed={this.state.animationSpeed} moveTile={this.moveTile.bind(this)} tilesData={this.state.tilesData} />
          </div>
        </div>
        <Menu shuffleClick={this.shuffleClick} changingColorsClick={this.changingColorsClick} changingColorsState={this.state.changingColors} />
      </div>

    );
  }
}

export default Game;
