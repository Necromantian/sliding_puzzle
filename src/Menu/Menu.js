import React, { Component } from 'react';
import './Menu.css';

class Menu extends Component {
  
  render() {
    let {changingColorsState} = this.props;
    return (
      <div className='menu'>
        <button onClick={this.props.shuffleClick} className='menu__button'>
          Shuffle Tiles
        </button>
        <button onClick={this.props.changingColorsClick} className={changingColorsState ? 'menu__button menu__button--clicked':'menu__button'}>
          Crazy Colors
        </button>
      </div>
    );
  }
}

export default Menu;