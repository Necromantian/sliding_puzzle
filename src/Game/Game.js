import React, { Component } from 'react';
import './Game.css';

import Tile from './Tile.js'

class Game extends Component {
  constructor(props) {
    super(props);
    this.state = {posX: 0, posY:0};
  };

  handleMouseDown = () => {
    this.setState({posX: this.state.posX+125});
  };

  handleMouseDown2 = () => {
    this.setState({posX: this.state.posX-125});
  };

  handleMouseDown3 = () => {
    this.setState({posY: this.state.posY-125});
  };

  handleMouseDown4 = () => {
    this.setState({posY: this.state.posY+125});
  };

  handleTouchStart = (e) => {
    e.preventDefault();
    this.handleMouseDown();
  };

  render() {
    return (
      <div className='boardContainer'>
        <div className="board">
        <button className='test'
          onMouseDown={this.handleMouseDown}>
          Toggle --
        </button>
        <button className='test2'
          onMouseDown={this.handleMouseDown2}>
          -- Toggle 
        </button>
        <button className='test3'
          onMouseDown={this.handleMouseDown3}>
          Toggle +
        </button>
        <button className='test4'
          onMouseDown={this.handleMouseDown4}>
         Toggle -
        </button>
          <Tile positionX={this.state.posX} positionY={this.state.posY}/>
        </div>
      </div>
    );
  }
}

export default Game;
