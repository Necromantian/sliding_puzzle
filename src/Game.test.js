import React from 'react';
import ReactDOM from 'react-dom';
import Game from './Game';
import Menu from './Menu/Menu';

import { mount, shallow, render } from 'enzyme';


describe('MyComponent', () => {
  it('should render correctly in "debug" mode', () => {
    const component = shallow(<Game />);
  
    expect(component).toMatchSnapshot();
  });
});

it('renders Game without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Game />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders Menu without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Menu />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('calc correctly', () => {
    test('15', () => {
        let  a = 5;
        let b = 10;
    
        expect(a+b).toBe(15);
    });
    test('11', () => {
        let  a = 1;
        let b = 10;
    
        expect(a+b).toBe(11);
    });
  });