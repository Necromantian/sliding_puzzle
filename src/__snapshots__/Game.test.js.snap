// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyComponent should render correctly in "debug" mode 1`] = `
<div
  className="game"
>
  <div
    className="boardContainer"
  >
    <div
      className="board"
    >
      <Board
        animationSpeed="slow"
        changingColors={false}
        moveTile={[Function]}
        tilesData={
          Array [
            Object {
              "currentPosition": 0,
              "number": 0,
            },
            Object {
              "currentPosition": 1,
              "number": 1,
            },
            Object {
              "currentPosition": 2,
              "number": 2,
            },
            Object {
              "currentPosition": 3,
              "number": 3,
            },
            Object {
              "currentPosition": 4,
              "number": 4,
            },
            Object {
              "currentPosition": 5,
              "number": 5,
            },
            Object {
              "currentPosition": 6,
              "number": 6,
            },
            Object {
              "currentPosition": 7,
              "number": 7,
            },
            Object {
              "currentPosition": 8,
              "number": 8,
            },
            Object {
              "currentPosition": 9,
              "number": 9,
            },
            Object {
              "currentPosition": 10,
              "number": 10,
            },
            Object {
              "currentPosition": 11,
              "number": 11,
            },
            Object {
              "currentPosition": 12,
              "number": 12,
            },
            Object {
              "currentPosition": 13,
              "number": 13,
            },
            Object {
              "currentPosition": 14,
              "number": 14,
            },
          ]
        }
      />
    </div>
  </div>
  <Menu
    changingColorsClick={[Function]}
    changingColorsState={false}
    shuffleClick={[Function]}
  />
</div>
`;
